### PR TITLE
ARROW-12197: [R] dplyr bindings for cast, dictionary_encode

### PR DESCRIPTION
**Pull request by ianmcook**
This adds bindings for `cast()`, `dictionary_encode()`, and `as.factor()` for dplyr ~~, and also adds some argument-defusing code that prevents other packages from masking the arrow data type functions (e.g. `rlang::string()`)~~

### CODE REVIEW

File: r/tests/testthat/test-dplyr.R [hunk 1]
```diff
@@ -421,12 +484,13 @@ test_that("explicit type conversions", {
...
-        lgl2chr = toupper(as.character(lgl)), # Arrow returns "true", "false"
+        lgl2chr = as.character(lgl), # Arrow returns "true", "false" here ...
         lgl2dbl = as.double(lgl),
         lgl2int = as.integer(lgl),
-        lgl2lgl = as.logical(lgl),
+        lgl2lgl = as.logical(lgl)
       ) %>%
-      collect(),
+      collect() %>%
+      mutate(lgl2chr = toupper(lgl2chr)), # ... but we need "TRUE", "FALSE"
```
Comment (ianmcook) [inline r/tests/testthat/test-dplyr.R:493]: I moved the `toupper()` to after `collect()` to allow this test to run successfully without utf8proc. This fixes a `test-r-minimal-build` failure

Comment (nealrichardson) [reply r/tests/testthat/test-dplyr.R:493]: Can you add a comment explaining that?

File: r/R/arrow-datum.R [hunk 2]
```diff
@@ -22,6 +22,7 @@
 ArrowDatum <- R6Class("ArrowDatum", inherit = ArrowObject,
   public = list(
     cast = function(target_type, safe = TRUE, ...) {
+      target_type <- unmask_type_fun(enexpr(target_type)) %||% target_type
```
Comment (nealrichardson) [inline r/R/arrow-datum.R:25]: I see you have this logic in `as_type` too; is that not enough? Seems like it would be best to encapsulate that complexity there somehow.

Comment (ianmcook) [reply r/R/arrow-datum.R:25]: Unfortunately for this to work, the `type`/`target_type` argument needs to be defused in the function the user passes it to, otherwise it will be evaluated there.

File: r/tests/testthat/test-dplyr.R (hunk 1 above)
Comment (ianmcook) [reply r/tests/testthat/test-dplyr.R:494]: Added in 95272bc268a5742ecf3a79828172f0de75801bc7

File: r/R/arrow-datum.R (hunk 2 above)
Comment (ianmcook) [reply r/R/arrow-datum.R:25]: P.S. my rationale for adding this masking-prevention code is that there are more than two dozen data type functions in the arrow package, and they are named very generically, so they represent a large surface area for potential masking problems, which are likely to occur in user environments, not in our CI. If these masking errors do occur, they will probably give frustratingly unhelpful error messages. This happened to me with `rlang::string()` as I was working on this PR. The error was
>Error in is_integerish(x) : argument "x" is missing, with no default

which confused the heck out of me for a couple minutes. I think the addition of this code should prevent the majority of such masking errors.

**Comment by ianmcook** [comment]
@nealrichardson this is ready for re-review. Thanks!

File: r/R/arrow-datum.R (hunk 2 above)
Comment (nealrichardson) [reply r/R/arrow-datum.R:25]: > Unfortunately for this to work, the `type`/`target_type` argument needs to be defused in the function the user passes it to, otherwise it will be evaluated there.

Isn't this what the `{{ }}` thing is used for?

File: r/R/type.R [hunk 3]
```diff
@@ -413,16 +413,28 @@ FixedSizeListType <- R6Class("FixedSizeListType",
...
-  if (identical(type, double())) {
-    # Magic so that we don't have to mask this base function
-    type <- float64()
-  }
+  # work around other packages possibly masking the Arrow data type functions,
+  # for example rlang masking string()
+  type <- unmask_type_fun(enexpr(type)) %||% type
+
+  # magic so we don't have to mask base::double()
+  if (identical(type, double())) type <- float64()
```
Comment (nealrichardson) [inline r/R/type.R:421]:
```suggestion
  if (identical(type, double())) {
    type <- float64()
  }
```

File: r/R/type.R [hunk 4]
```diff
@@ -156,8 +156,8 @@ NestedType <- R6Class("NestedType", inherit = DataType)
 #' * `float16()` and `halffloat()`
 #' * `float32()` and `float()`
 #' * `bool()` and `boolean()`
-#' * Called from `schema()` or `struct()`, `double()` also is supported as a
-#' way of creating a `float64()`
+#' * Called from `schema()` or `struct()` or from `cast()` in a dplyr verb,
```
Comment (nealrichardson) [inline r/R/type.R:159]:
```suggestion
#' * When called inside an `arrow` function, such as `schema()` or `cast()`,
```

File: r/tests/testthat/test-dplyr.R [hunk 5]
```diff
@@ -358,7 +358,70 @@ test_that("relocate with selection helpers", {
...
+  # Arrow errors when truncating floats...
+  expect_error(
+    expect_type_equal(
+      {
+        t1 <- Table$create(pi = pi) %>%
+          transmute(three = cast(pi, int32())) %>%
+          compute()
+        t1$schema[[1]]$type
+      },
+      as_type(int32())
```
Comment (nealrichardson) [inline r/tests/testthat/test-dplyr.R:407]: `int32()` is a type so you don't need `as_type()` do you? (I'm guessing this is copypasta from the above tests, where you do need it to handle `double()`)

```suggestion
      int32()
```

File: r/R/arrow-datum.R (hunk 2 above)
Comment (ianmcook) [reply r/R/arrow-datum.R:25]: `{{x}}` is shorthand for `!!enquo(x)` which creates a quosure then evaluates it. I don't think that works in this context. I tried some various other approaches, e.g. just using `enoquo(x)`, using `eval(substitute(x))`, etc. (which some of my earlier commits show the history of) but those failed to handle all cases. In particular, we want to handle all these cases:
- `type` is a variable in the calling environment whose value is a `DataType` object
-  type` is a call to a function defined by the user that returns a `DataType` object
- `type` is any arbitrary R expression that the user has wired up to return a `DataType` object
- `type` is a call to a function in another package or in the user's environment that masks the Arrow type function of the same name

The `%||% type` part of in `unmask_type_fun(enexpr(type)) %||% type` is basically saying "if you tried to defuse `type` and evaluate in the arrow package environment but that failed, then give up on the NSE approach and just evaluate `type` in the standard way. That was important for enabling the first three cases above to work.

Comment (ianmcook) [reply r/R/arrow-datum.R:25]: Re the `%||% type` thing: To create a foolproof escape hatch for when the NSE approach fails to evaluate `type` properly, you need to evaluate `type` in the called function itself, not in another function called from that function.

Comment (ianmcook) [reply r/R/arrow-datum.R:25]: P.S. If every call to `as_type()` happened at a fixed number of function calls `n` from the calling environment, then we could simply evaluate the defused expression `type` there, using `eval.parent(type, n)`. That would allow us to encapsulate all the NSE evaluation in `as_type()`. But that's not the case.

Comment (ianmcook) [reply r/R/arrow-datum.R:25]: Oh, I take that back, there is a sneaky way to do this while encapsulating all this logic in `as_type()`. But I'm not sure it's really cleaner and better. I'll describe it here in a moment.

Comment (ianmcook) [reply r/R/arrow-datum.R:25]: Using code like this in `unmask_type_fun` would avoid the need for `unmask_type_fun` to be called anywhere except in `as_type()`:
```r
i <- 1L
while (identical(utils::packageName(parent.frame(i)), "arrow")) {
  i <- i + 1L
}
return(eval.parent(expr, n = i))
```
Then all we would need in the other functions is to wrap `type` in `enexpr()`.

I'm on the fence about whether this is an improvement (because encapsulation) or not (because hard to follow).

Comment (ianmcook) [reply r/R/arrow-datum.R]: Ugh—this also causes problems with tests because testthat runs within the package environment.

Comment (ianmcook) [reply r/R/arrow-datum.R]: I think there's a way to solve this with quosures.

Comment (ianmcook) [reply r/R/arrow-datum.R]: My first attempt to solve this with quosures was unsuccessful for two reasons:
- There are checks for the nullity of `type` in the code would need to all be rewritten to work with quosures, and simply using `quo_is_null()` is not sufficient for this.
- Some of the functions that defuse `type` call other functions that defuse `type`, so we end up with quosures of quosures. rlang does not appear to have facilities for easily solving this problem. See https://community.rstudio.com/t/how-to-avoid-creating-quosures-of-quosures/101420

I will remove all the masking-prevention code for now and we can revisit in ARROW-12322.

**Review by nealrichardson — APPROVED** [review]

File: r/R/arrow-datum.R (hunk 2 above)
Comment (ianmcook) [reply r/R/arrow-datum.R]: Re the nullity checks mentioned in the above comment: we can probably get around that problem by using `if (!missing(type))` instead of `if (!is.null(type))` and not giving the `type` argument a default value of `NULL`. This solves two problems: 1- checking `is.null()` on a quosure causes it to evaluate immediately and makes subsequent calls to `enquo()` capture the evaluated value, not the expression. 2- checking `is.null()` on a quosure whose expression evaluates to `NULL` does not return `NULL`.
